### PR TITLE
Repair CreateDepartmentCommandTest

### PR DESCRIPTION
**Pull request by Cheephot**

### CODE REVIEW

File: src/CompanyService.Business/Commands/Department/CreateDepartmentCommand.cs [hunk 1]
```diff
@@ -41,25 +41,25 @@ public class CreateDepartmentCommand : ICreateDepartmentCommand
 
         public OperationResultResponse<Guid> Execute(CreateDepartmentRequest request)
         {
-            if (!(_accessValidator.IsAdmin() || _accessValidator.HasRights(Rights.AddEditRemoveDepartments)))
+            if (!(_accessValidator.IsAdmin() || _accessValidator.HasRights(Rights.AddEditRemoveDepartments)))
```
Comment (BelehovEgor) [inline src/CompanyService.Business/Commands/Department/CreateDepartmentCommand.cs:44]: remove added spaces

File: test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs [hunk 2]
```diff
@@ -77,108 +66,186 @@ public void OneTimeSetUp()
...
-                .Setup(x => x.HasRights(Rights.AddEditRemoveDepartments))
+            _autoMock
+                .Setup<IAccessValidator, bool>(x => x.HasRights(Rights.AddEditRemoveDepartments))
                 .Returns(false);
 
-            Assert.Throws<ForbiddenException>(() => _command.Execute(_request));
-            _companyRepositoryMock.Verify(x => x.Get(It.IsAny<GetCompanyFilter>()), Times.Never);
-            _repositoryMock.Verify(x => x.CreateDepartment(It.IsAny<DbDepartment>()), Times.Never);
+            var ex = Assert.Throws<ForbiddenException>(() => _command.Execute(_request));
+            Assert.That(ex.Message, Is.EqualTo("Not enough rights."));
```
Comment (BelehovEgor) [inline test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs:118]: use Assert.Throws

File: test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs [hunk 3]
```diff
@@ -77,108 +66,186 @@ public void OneTimeSetUp()
...
-            _repositoryMock.Verify(x => x.CreateDepartment(_dbDepartment), Times.Never);
-        }*/
+            _autoMock.Verify<IDepartmentRepository, Guid>(
+                x => x.CreateDepartment(It.IsAny<DbDepartment>()), 
+                Times.Never);
+        }
 
         [Test]
-        public void ShouldAddNewDepartmentSuccessfully()
+        public void CompanyDoesNotExistTest()
```
Comment (BelehovEgor) [inline test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs:149]: rename

Comment (BelehovEgor) [reply test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs:149]: should throw exc when company doesn't exist

File: test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs [hunk 4]
```diff
@@ -77,108 +66,186 @@ public void OneTimeSetUp()
...
+                x => x.Get(It.IsAny<GetCompanyFilter>()), 
+                Times.Once);
+
+            _autoMock.Verify<IDepartmentRepository, Guid>(
+                x => x.CreateDepartment(It.IsAny<DbDepartment>()), 
+                Times.Never);
+        }
+
+        [Test]
+        public void TheDepartmentNameAlreadyExistsTest()
```
Comment (BelehovEgor) [inline test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs:172]: rename

File: src/CompanyService.Business/Commands/Department/CreateDepartmentCommand.cs (hunk 1 above)
Comment (Cheephot) [reply src/CompanyService.Business/Commands/Department/CreateDepartmentCommand.cs]: Done

File: test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs (hunk 2 above)
Comment (Cheephot) [reply test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs]: Done

File: test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs (hunk 3 above)
Comment (Cheephot) [reply test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs]: Done

File: test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs (hunk 4 above)
Comment (Cheephot) [reply test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs]: Done

File: test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs [hunk 5]
```diff
@@ -77,108 +66,185 @@ public void OneTimeSetUp()
...
         }
 
-        /*[Test]
-        public void ShouldThrowArgumentNullExceptionWhenDepartmentRequestIsNull()
+        [Test]
+        public void ShouldThrowValidationException()
         {
+            _autoMock
+                .Setup<ICreateDepartmentRequestValidator, bool>
+                (x => x.Validate(It.IsAny<IValidationContext>()).IsValid)
```
Comment (VLiamin) [inline test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs:133]: Tab?

Comment (Cheephot) [reply test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs:133]: Not needed

File: test/CompanyService.Business.UnitTests/CompanyService.Business.UnitTests.csproj [hunk 6]
```diff
@@ -2,6 +2,10 @@
 
   <Import Project="$(ProjectDir)..\..\Common.Configurations.targets" />
 
+  <ItemGroup>
```
Comment (BelehovEgor) [inline test/CompanyService.Business.UnitTests/CompanyService.Business.UnitTests.csproj:5]: what is this?

File: test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs [hunk 7]
```diff
@@ -77,108 +66,184 @@ public void OneTimeSetUp()
...
+            _request = GenerateRequest(_directorId);
+
+            Assert.DoesNotThrow(() => _command.Execute(_request));
+
+            _autoMock.Verify<IDepartmentUserRepository>(
+                x => x.Remove(_directorId),
+                Times.Once);
+
+            _autoMock.Verify<IDepartmentUserRepository>(
+            x => x.Remove(It.IsAny<Guid>()),
```
Comment (BelehovEgor) [inline test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs:225]: tab

File: test/CompanyService.Business.UnitTests/CompanyService.Business.UnitTests.csproj (hunk 6 above)
Comment (Cheephot) [reply test/CompanyService.Business.UnitTests/CompanyService.Business.UnitTests.csproj:5]: NuGet AutoMoq, i delete this

File: test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs [hunk 8]
```diff
@@ -77,108 +66,184 @@ public void OneTimeSetUp()
...
-            _repositoryMock
-                .Setup(x => x.CreateDepartment(_dbDepartment))
-                .Returns(_dbDepartment.Id);
+            _autoMock.Verify<IDepartmentRepository, Guid>(
+                x => x.CreateDepartment(It.IsAny<DbDepartment>()),
+                Times.Never);
+        }
+
+        [Test]
+        public void RemoveUserTest()
```
[the remark this answers is not in the document]
Comment (Cheephot) [reply test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs:194]: done

**Review by ndoriomedov — APPROVED** [review]

File: test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs [hunk 9]
```diff
@@ -22,51 +23,39 @@ namespace LT.DigitalOffice.CompanyService.Business.UnitTests.Commands.Department
...
-        private Mock<ICompanyRepository> _companyRepositoryMock;
-        private Mock<IAccessValidator> _accessValidatorMock;
-        private Mock<ICreateDepartmentRequestValidator> _validatorMock;
-        private Mock<IDbDepartmentMapper> _mapperMock;
+        private AutoMocker _autoMock;
 
         private CreateDepartmentRequest _request;
         private DbDepartment _dbDepartment;
 
+        private readonly Guid _directorId = new Guid("528E413B-7A3D-4CA9-851A-B9DA5840B216");
```
Comment (MariyaSin) [inline test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs:31]: Why you write guid like this? You can write Guid.NewGuid

File: test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs [hunk 10]
```diff
@@ -77,108 +66,184 @@ public void OneTimeSetUp()
...
         {
-            _repositoryMock.Reset();
-            _accessValidatorMock.Reset();
-            _validatorMock.Reset();
-            _mapperMock.Reset();
-            _companyRepositoryMock.Reset();
-
-            _accessValidatorMock
-                .Setup(x => x.IsAdmin(null))
+            _autoMock.GetMock<IDepartmentRepository>().Reset();
```
Comment (MariyaSin) [inline test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs:74]: Why do you do reset?

File: test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs (hunk 9 above)
Comment (Cheephot) [reply test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs:31]: Done

**Review by VLiamin — APPROVED** [review]

File: test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs [hunk 11]
```diff
@@ -89,108 +66,188 @@ public void OneTimeSetUp()
...
-            _companyRepositoryMock.Verify(x => x.Get(It.IsAny<GetCompanyFilter>()), Times.Never);
-            _repositoryMock.Verify(x => x.CreateDepartment(It.IsAny<DbDepartment>()), Times.Never);
+
+            _autoMock.Verify<ICompanyRepository, DbCompany>(
+                x => x.Get(It.IsAny<GetCompanyFilter>()),
+                Times.Never);
+
+            _autoMock.Verify<IDepartmentRepository, Guid>(
+                x => x.CreateDepartment(It.IsAny<DbDepartment>()),
+                Times.Never);
```
Comment (MariyaSin) [inline test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs:129]: you can add verify for DoesNameExist and mapper

File: test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs [hunk 12]
```diff
@@ -89,108 +66,188 @@ public void OneTimeSetUp()
...
+
+            Assert.Throws<ValidationException>(() => _command.Execute(_request));
+
+            _autoMock.Verify<ICompanyRepository, DbCompany>(
+                x => x.Get(It.IsAny<GetCompanyFilter>()),
+                Times.Never);
+
+            _autoMock.Verify<IDepartmentRepository, Guid>(
+                x => x.CreateDepartment(It.IsAny<DbDepartment>()),
+                Times.Never);
```
Comment (MariyaSin) [inline test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs:147]: you can add verify for DoesNameExist and mapper

File: test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs [hunk 13]
```diff
@@ -89,108 +66,188 @@ public void OneTimeSetUp()
...
         }
 
-        /*[Test]
-        public void ShouldThrowArgumentNullExceptionWhenDepartmentRequestIsNull()
+        [Test]
+        public void ShouldThrowExcWhenCompanyDoesntExist()
```
Comment (MariyaSin) [inline test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs:151]: ShouldThrowExcWhenCompanyDoesntExist => ShouldThrowExceptionWhenCompanyNameAlreadyExist

File: test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs [hunk 14]
```diff
@@ -89,108 +66,188 @@ public void OneTimeSetUp()
...
+                x => x.Get(It.IsAny<GetCompanyFilter>()),
+                Times.Once);
 
-            Assert.Throws<ArgumentNullException>(() => _command.Execute(null));
-            _companyRepositoryMock.Verify(x => x.Get(null), Times.Once);
-            _repositoryMock.Verify(x => x.CreateDepartment(_dbDepartment), Times.Never);
-        }*/
+            _autoMock.Verify<IDepartmentRepository, Guid>(
+                x => x.CreateDepartment(It.IsAny<DbDepartment>()),
+                Times.Never);
```
Comment (MariyaSin) [inline test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs:170]: you can add verify for DoesNameExist and mapper

File: test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs [hunk 15]
```diff
@@ -89,108 +66,188 @@ public void OneTimeSetUp()
...
+        }
 
         [Test]
-        public void ShouldAddNewDepartmentSuccessfully()
+        public void ShouldThrowExcWhenDepartmentNameAlreadyExists()
```
Comment (MariyaSin) [inline test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs:174]: please do not short the word Exception

Comment (Cheephot) [reply test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs]: done

File: test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs (hunk 14 above)
Comment (Cheephot) [reply test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs:196]: done

File: test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs (hunk 13 above)
Comment (Cheephot) [reply test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs]: done

File: test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs (hunk 12 above)
Comment (Cheephot) [reply test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs:160]: done

File: test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs (hunk 11 above)
Comment (Cheephot) [reply test/CompanyService.Business.UnitTests/Commands/Department/CreateDepartmentCommandTests.cs:130]: done

**Review by borovHokage — APPROVED** [review]

**Pull request merged**